p2panda-net/src/sync/accept.rs

```diff
@@ -0,0 +1,221 @@
+    let (tx, mut rx) = mpsc::channel::<FromSync<T>>(128);
+    let mut sink = PollSender::new(tx).sink_map_err(|e| SyncError::Critical(e.to_string()));
+
+    // Set up a channel for sending over errors to the "glue" task which occurred during sync.
+    let (sync_error_tx, mut sync_error_rx) = oneshot::channel::<SyncError>();
+
+    // Spawn a "glue" task which represents the layer between the sync session and the engine.
+    //
+    // It picks up any messages from the sync session makes sure that the "Two-Phase Sync Flow" is
+    // followed (I. "Handshake" Phase & II. "Data Sync" Phase) and the engine accordingly informed
+    // about it.
+    //
+    // If the task detects any invalid behaviour of the sync flow, it fails critically, indicating
+    // that the sync protocol implementation does not behave correctly and is not compatible with
+    // the engine.
+    //
+    // Additionally, the task forwards any synced application data straight to the engine.
+    let glue_task_handle: JoinHandle<Result<(), SyncError>> = tokio::spawn(async move {
+        let mut topic = None;
+
+        loop {
+            tokio::select! {
+                biased;
+
+                Ok(err) = &mut sync_error_rx => {
+                    engine_actor_tx
+                        .send(ToEngineActor::SyncFailed {
+                            peer,
+                            topic: topic.clone(),
+                        })
+                        .await
+                        .map_err(|err| {
+                            SyncError::Critical(
+                                format!("engine_actor_tx failed sending sync failed: {err}")
+                            )
+                        })?;
+
+                    // If we're observing an error we terminate the task here and propagate that
+                    // error further up.
+                    return Err(err);
+                },
+                message = rx.recv() => {
+                    let Some(message) = message else {
+                        // Sink (tx) got dropped, so we're leaving the task.
+                        break;
+                    };
+
+                    // I. Handshake Phase.
+                    //
+                    // At the beginning of every sync session the "accepting" peer needs to learn
+                    // the topic of the "initiating" peer during the handshake phase. This is
+                    // _always_ the first message we're expecting:
+                    if let FromSync::HandshakeSuccess(handshake_topic) = message {
+                        // It should only be sent once so topic should be `None` now.
+                        if topic.is_some() {
+                            return Err(
+                                SyncError::Critical(
+                                    "received topic twice from sync session in handshake phase"
+                                    .into()
+                                )
```

Comment:
All violations of the "2 Phase Protocol Flow" are considered critical errors (as we can assume they are due to buggy `Sync` implementations.